### PR TITLE
mute unnecessary logs when failing to parse IPs

### LINKED ISSUE #105182: Constant 'Skipping invalid IP: ".' error entries in Kube-Proxy logs
#### What happened:

As we update Kube-proxy to version 1.21, we observe a stream of following error entries within the Kube-proxy logs:

Skipping invalid IP: ".

#### What you expected to happen:

As we do not see anything wrong with the EKS cluster itself, this message should not be reported as an error.

#### How to reproduce it (as minimally and precisely as possible):

1. Launch an EKS cluster v1.21 and launched a managed nodegroup within this cluster.

2. Either use the managed add-on or manage Kube-proxy yourself. Either way, the EKS provided kube-proxy image (kube-proxy:v1.21.2-eksbuild.2) is based on the upstream image and this error can be traced to the upstream code as I will be outlining in the next section.

3. Check logs for the kube-proxy pod:

kubectl logs -n kube-system <kube-proxy-pod-name>

#### Anything else we need to know?:

In my replication, I could analyze this:

These errors were recurring in the kube-proxy logs every 15 minutes which I concluded after watching the log stream for a 1+ hour window.

To be able to make sense of the logs better, I increased verbosity of kubeproxy to 9:

kubectl patch ds kube-proxy -n kube-system --type='json' -p='[{"op": "replace", "path": "/spec/template/spec/containers/0/command/1", "value":"--v=9"}]'

In the logs, I could notice the following snippet recurring periodically (every 15 minutes):

```
I0921 17:44:05.867454       1 reflector.go:381] k8s.io/client-go/informers/factory.go:134: forcing resync
I0921 17:44:05.867508       1 reflector.go:381] k8s.io/client-go/informers/factory.go:134: forcing resync
I0921 17:44:05.867605       1 config.go:351] Calling handler.OnServiceUpdate
I0921 17:44:05.867710       1 config.go:351] Calling handler.OnServiceUpdate
I0921 17:44:05.867742       1 config.go:351] Calling handler.OnServiceUpdate
E0921 17:44:05.867754       1 utils.go:282] Skipping invalid IP: 
E0921 17:44:05.867766       1 utils.go:282] Skipping invalid IP: 
I0921 17:44:05.867792       1 config.go:260] Calling handler.OnEndpointSliceUpdate
I0921 17:44:05.867835       1 config.go:260] Calling handler.OnEndpointSliceUpdate
```

This hinted that a 'resync' was behind causing this errors being pushed every 15 minutes, however, before diving deeper into the root cause there, I searched for the origin of 'Skipping invalid IP: ' line and tried to reverse engineer back to the root cause for these errors.

For this, I went through the 'pkg/proxy/util/utils.go' in kube-proxy upstream github (release-1.21 branch). 

I started with the 'MapIPsByIPFamily' function which is where this error is handled and thrown [1]. This error is returned to it by a child function getIPFamilyFromIP [2].

The purpose of this function is to identify if the IP string passed is IPv6 or IPv4.

So, I backtracked a bit from the 'MapIPsByIPFamily' function and observed that it was referenced on line 186 by 'newBaseServiceInfo' in the service.go file of kube-proxy [3]. Here I noticed that 'service.Spec.ExternalIPs' is passed as string to the 'MapIPsByIPFamily' function. 

Now, not all services have an ExternalIP defined in their spec. This is when an empty string '' is passed to the latter and hence the snippet : [klog.Errorf("Skipping invalid IP: %s", ip)] translates to 'Skipping invalid IP: ' in the actual error message.

However, in the following error handling block:

```
for _, ip := range ipStrings {
		// Handle only the valid IPs
		if ipFamily, err := getIPFamilyFromIP(ip); err == nil {
			ipFamilyMap[ipFamily] = append(ipFamilyMap[ipFamily], ip)
		} else {
			klog.Errorf("Skipping invalid IP: %s", ip)
		}
	}

```
Rather than simply logging an error, there could be a check for this case of an empty string (as this is pretty common for clusters to have services without ExternalIP defined). If it is not possible to handle this error in this snippet for some reason, the verbosity level could be highered to atleast 4 so that the default kube-proxy image that runs at level 2 generally is not bombarded with a stream of recurring errors that are not actually errors.

Can you please help fix this issue?

[1] https://github.com/kubernetes/kubernetes/blob/release-1.21/pkg/proxy/util/utils.go#L275

[2] https://github.com/kubernetes/kubernetes/blob/release-1.21/pkg/proxy/util/utils.go#L302

[3] https://github.com/kubernetes/kubernetes/blob/release-1.21/pkg/proxy/service.go#L186

#### Environment:
- Kubernetes version (use `kubectl version`): 1.21
- Cloud provider or hardware configuration: EKS with EC2 worker nodes
- OS (e.g: `cat /etc/os-release`): Amazon Linux 2
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by khenidak**
#### What type of PR is this?
/kind bug
/kind cleanup

#### What this PR does / why we need it:
Fixes https://github.com/kubernetes/kubernetes/issues/105182

#### Which issue(s) this PR fixes:
Fixes #105182

#### Special notes for your reviewer:
N/A 

#### Does this PR introduce a user-facing change?
```release-note
NONE
```
/sig network
/assign @aojea

### CODE REVIEW

File: pkg/proxy/util/utils.go [hunk 1]
```diff
@@ -288,7 +288,14 @@ func MapCIDRsByIPFamily(cidrStrings []string) map[v1.IPFamily][]string {
...
 			ipFamilyMap[ipFamily] = append(ipFamilyMap[ipFamily], cidr)
 		} else {
-			klog.ErrorS(nil, "Skipping invalid CIDR", "cidr", cidr)
+			// this function is called in multiple places. All of which
+			// have sanitized data. Except the case of ExternalIPs which is
+			// not validated by api-server. Specifically empty strings
+			// validation. Which yields into a lot of bad error logs.
+			// check for empty string
+			if cidr != "" && cidr != " " {
+				klog.ErrorS(nil, "Skipping invalid CIDR", "cidr", cidr)
```
Comment (aojea) [inline pkg/proxy/util/utils.go:297]: the bug is in the previous function
```
Skipping invalid IP: ".
```

this only used for nodePortAddresses and loadBalancerSourceRanges
```
pkg/proxy/iptables/proxier.go:  ipFamilyMap := utilproxy.MapCIDRsByIPFamily(nodePortAddresses)
pkg/proxy/iptables/proxier.go:  ipFamilyMap := utilproxy.MapCIDRsByIPFamily(nodePortAddresses)
pkg/proxy/ipvs/proxier.go:      ipFamilyMap := utilproxy.MapCIDRsByIPFamily(nodePortAddresses)
pkg/proxy/ipvs/proxier.go:      ipFamilyMap := utilproxy.MapCIDRsByIPFamily(nodePortAddresses)
pkg/proxy/util/utils_test.go:func TestMapCIDRsByIPFamily(t *testing.T) {
pkg/proxy/util/utils_test.go:                   cidrMap := MapCIDRsByIPFamily(testcase.ipString)
pkg/proxy/util/utils.go:// MapCIDRsByIPFamily maps a slice of IPs to their respective IP families (v4 or v6)
pkg/proxy/util/utils.go:func MapCIDRsByIPFamily(cidrStrings []string) map[v1.IPFamily][]string {
pkg/proxy/service.go:   ipFamilyMap = utilproxy.MapCIDRsByIPFamily(loadBalancerSourceRanges)
```

loadbalancerSourceRanges had the same issue and was fixed with this https://github.com/kubernetes/kubernetes/pull/94107

File: pkg/proxy/util/utils.go [hunk 2]
```diff
@@ -288,7 +288,14 @@ func MapCIDRsByIPFamily(cidrStrings []string) map[v1.IPFamily][]string {
 		if ipFamily, err := getIPFamilyFromCIDR(cidr); err == nil {
 			ipFamilyMap[ipFamily] = append(ipFamilyMap[ipFamily], cidr)
 		} else {
-			klog.ErrorS(nil, "Skipping invalid CIDR", "cidr", cidr)
+			// this function is called in multiple places. All of which
+			// have sanitized data. Except the case of ExternalIPs which is
+			// not validated by api-server. Specifically empty strings
+			// validation. Which yields into a lot of bad error logs.
+			// check for empty string
+			if cidr != "" && cidr != " " {
```
Comment (aojea) [inline pkg/proxy/util/utils.go:296]: if len(strings.TrimSpace(ip)) == 0 {
   continue
}

File: pkg/proxy/util/utils.go [hunk 3]
```diff
@@ -274,7 +275,15 @@ func MapIPsByIPFamily(ipStrings []string) map[v1.IPFamily][]string {
 		if ipFamily, err := getIPFamilyFromIP(ip); err == nil {
 			ipFamilyMap[ipFamily] = append(ipFamilyMap[ipFamily], ip)
 		} else {
-			klog.ErrorS(nil, "Skipping invalid IP", "ip", ip)
+			// this function is called in multiple places. All of which
+			// have sanitized data. Except the case of ExternalIPs which is
+			// not validated by api-server. Specifically empty strings
+			// validation. Which yields into a lot of bad error logs.
+			// check for empty string
+			if len(strings.TrimSpace(ip)) == 0 {
```
Comment (aojea) [inline pkg/proxy/util/utils.go:283]:
```suggestion
			if len(strings.TrimSpace(ip)) != 0 {
```

:smile:

**Comment by aojea** [comment]
/lgtm
/approve 
Thanks for taking this

**Comment by hakman** [comment]
@aojea Any thoughts on cherry-picking this to the supported versions? Would create less panic for operators. 😄

**Pull request merged**